doc/config.md

```diff
@@ -300,6 +300,79 @@ updated and how often aggregated statistics are written to the log
 
 Default: 60
 
+### prepared_statement_cache_size
+
+When this is set to a non-zero value PgBouncer will track protocol-level named
+prepared statements related commands sent by the client in transaction and
+statement pooling mode. PgBouncer will make sure that any statement prepared by
+a client is available on the backing server connection. Even when the statement
+was originally prepared on another server connection.
+
+The way this prepared statement support works is that PgBouncer will internally
+examine all the queries that are sent as a prepared statement by clients and
+will give each unique query string an internal name with the format
+`PGBOUNCER_{unique_id}`. Prepared statements will only be prepared using this
+name on the actual PostgreSQL server. Then for each client PgBouncer keeps
+track of the name that the client gave to each prepared statement. It will then
+rewrite each command that uses a prepared statement to use the matching
+internal name (e.g. `PGBOUNCER_123`) before forwarding that command to the
+server. And importantly, if the prepared statement that the client wants to use
+is not prepared on the server yet, it will automatically prepare that statement
+before forwarding the command that the client sent.
+
+Note: This tracking and rerwriting of prepared statement commands does not work
+for SQL-level prepared statement commands such as `PREPARE`, `EXECUTE`,
+`DEALLOCATE`, `DEALLOCATE ALL` and `DISCARD ALL`. Running `DEALLOCATE ALL`
+and `DISCARD ALL` is especially problematic, since those commands will
+apear to run successfully, but they will also mess with state of the server
+connection significantly, without PgBouncer noticing. Which in turn will
+very likely break the execution of any further prepared statements on that
+server connection.
+
+The actual value of this setting controls the number of prepared statements
+kept active on a single server connection. When the setting is set to 0
+prepared statement support for transaction and statement pooling is disabled.
+Keep in mind that the higher this value, the larger the memory footprint of
+each PgBouncer connection will be on your PostgreSQL server. It also increases
+the memory footprint of PgBouncer itself, because it now needs to keep track of
+query strings.
+
+The impact on PgBouncer its memory usage is not that big though:
+- Each unique query is stored once in a global query cache.
```

Comment:
@knizhnik @alphavector moving the discussion about the `cached plan must not change result type` error here to not clutter the main thread too much. As @knizhnik described well, this is the issue in question:

```
postgres=# create table t(x integer);
postgres=# insert into t values (1);
postgres=# prepare ps1 as select * from t;
postgres=# alter table t alter column x type bigint;
postgres=# execute ps1;
ERROR:  cached plan must not change result type
postgres=# prepare ps2 as select count(*) from t where x > 0;
postgres=# alter table t alter column x type integer;
postgres=# execute ps2;
 count 
-------
     1
(1 row)
```

> > Any alter column type violates the prepared statements
> 
> > Sounds like you found a bug in Postgres. According to the docs any DDL on the referenced objects should invalidate the prepared statements and cause a replanning.
> 
> Both statements are not true. Postgres really performs invalidation of prepared statements affected by DDLs.
But it prohibits to change result type of prepared statements:

Honestly this still feels like a bug in Postgres to me. I can't find it being documented anywhere that changing the return type of a prepared statement breaks the invalidation that Postgres documents that it does (except for people on the internet getting the same error). I'll probably take a look at the postgres code in the next few weeks to find out how easy this is to fix.

> There is one more problem with this approach which may be even more serious than DDL problem: statements can be the same, but compiled in different context (for example `search_path`). There may be multiple schemas containing table with the same name. First session prepare statement using one schema, and another session - using second schema. So the same query test produce two different plans. Right now pgbouncer will not be able to detect it.
But IMHO this is "shut you own leg" scenario and at is not so important to address this issue.

For completeness, here is an example of this issue:
```sql
create schema int_return;
create schema bigint_return;
create table int_return.t(a int);
create table bigint_return.t(a bigint);
set search_path = int_return;
prepare ps as select * from t;
execute ps;
set search_path = bigint_return;
execute ps;
-- ERROR:  0A000: cached plan must not change result type
```

I agree that this case seems rather unlikely to actually happen to people.

> Certainly it may be possible to implement both approaches and let user choose. But it increase complexity almost twice. I am not sure that we should do it. 

I agree that implementing both and allow people to choose seems unwanted for this version. The patch is already getting large enough and this is really an edge case. If someone runs into this due to a migration that changes a column type running `RECONNECT` should make the issue go away, as it will clear all the server connections. I think that's an acceptable workaround.

Reply:
FYI I created a patch to try and fix this error in Postgres: https://commitfest.postgresql.org/44/4518/

Reply:
Wow, you did that pretty promptly.

I get an error on the simplest `select 1`

```
2023-08-26 21:58:08.376 UTC [60] LOG: execute __asyncpg_stmt_5__: show transaction isolation level
2023-08-26 21:58:08.376 UTC [60] ERROR: unsupported format code: 28271
2023-08-26 21:58:08.376 UTC [60] STATEMENT: show transaction isolation level
```

I tried versions 14.9 and 16beta3 with your patches

According to the documentation `show` cannot be specified in prepare, but apparently through a binary I/O protocol that uses asyncpg it is possible

Reply:
Sorry, I was not able to reproduce the problem.
Are you using the same python script as in https://github.com/alphavector/docker-pgbouncer
I switched on logging of all statements in Postgres and see that `show transaction_isolation_level` was actually executed:
```
2023-08-27 22:11:30.482 EEST [348072] LOG:  statement: SET client_encoding='''utf-8''';
2023-08-27 22:11:30.488 EEST [348072] LOG:  execute PGBOUNCER_0: SELECT
	    t.oid,
	    t.typelem     AS elemtype,
	    t.typtype     AS kind
	FROM
	    pg_catalog.pg_type AS t
	WHERE
	    t.oid = $1
	
2023-08-27 22:11:30.488 EEST [348072] DETAIL:  parameters: $1 = '114'
2023-08-27 22:11:30.489 EEST [348072] LOG:  statement: SET client_encoding='''utf-8''';
2023-08-27 22:11:30.489 EEST [348072] LOG:  statement: SET client_encoding='''utf-8''';
2023-08-27 22:11:30.489 EEST [348072] LOG:  execute PGBOUNCER_0: SELECT
	    t.oid,
	    t.typelem     AS elemtype,
	    t.typtype     AS kind
	FROM
	    pg_catalog.pg_type AS t
	WHERE
	    t.oid = $1
	
2023-08-27 22:11:30.489 EEST [348072] DETAIL:  parameters: $1 = '3802'
2023-08-27 22:11:30.490 EEST [348072] LOG:  statement: SET client_encoding='''utf-8''';
2023-08-27 22:11:30.490 EEST [348072] LOG:  statement: BEGIN ISOLATION LEVEL READ COMMITTED;
2023-08-27 22:11:30.491 EEST [348072] LOG:  execute PGBOUNCER_1: select pg_catalog.version()
2023-08-27 22:11:30.493 EEST [348072] LOG:  execute PGBOUNCER_2: select current_schema()
2023-08-27 22:11:30.494 EEST [348072] LOG:  execute PGBOUNCER_3: show transaction isolation level
2023-08-27 22:11:30.495 EEST [348072] LOG:  execute PGBOUNCER_4: show standard_conforming_strings
2023-08-27 22:11:30.495 EEST [348072] LOG:  statement: ROLLBACK;
2023-08-27 22:11:30.497 EEST [348072] LOG:  statement: SET client_encoding='''utf-8''';
2023-08-27 22:11:30.497 EEST [348072] LOG:  statement: BEGIN ISOLATION LEVEL READ COMMITTED;
2023-08-27 22:11:30.498 EEST [348072] LOG:  execute PGBOUNCER_5: SELECT 1
2023-08-27 22:11:30.498 EEST [348072] LOG:  statement: COMMIT;
```

Reply:
yes, I didnt push the actual changes, I apologize
you can run it right in the master branch
you don't even have to up pgbouncer images
```
docker compose up --build -d db worker
docker exec -it worker python -m select_one -D postgresql+asyncpg://test:test@db:5432/test
```

Reply:
As an image, I totally took
https://github.com/docker-library/postgres/blob/master/16/alpine3.18/Dockerfile

I just added a couple lines:
```
curl https://www.postgresql.org/message-id/attachment/149655/v1-0001-Support-prepared-statement-invalidation-when-resu.patch | patch -p1
curl https://www.postgresql.org/message-id/attachment/149656/v1-0002-Completely-remove-fixed_result-from-CachedPlanSou.patch | patch -p1
```


There may be a problem with the difference between applying `patch` and `git patch`, I'm still looking into this

Reply:
I still not able to reproduce the porblem. Both simple.py an one_select.py are working correctly:
{{{
$ python -m select_one -D postgresql+asyncpg://localhost:5433/postgres
2023-08-28 14:37:10,337 INFO sqlalchemy.engine.Engine select pg_catalog.version()
2023-08-28 14:37:10,337 INFO sqlalchemy.engine.Engine [raw sql] ()
2023-08-28 14:37:10,338 INFO sqlalchemy.engine.Engine select current_schema()
2023-08-28 14:37:10,339 INFO sqlalchemy.engine.Engine [raw sql] ()
2023-08-28 14:37:10,339 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2023-08-28 14:37:10,339 INFO sqlalchemy.engine.Engine [raw sql] ()
2023-08-28 14:37:10,340 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-08-28 14:37:10,341 INFO sqlalchemy.engine.Engine SELECT 1
2023-08-28 14:37:10,341 INFO sqlalchemy.engine.Engine [generated in 0.00011s] ()
2023-08-28 14:37:10,342 INFO sqlalchemy.engine.Engine COMMIT
}}}
I have tested only most recent version of postgresql.

Reply:
This error is actually caused by this patches.
The problem is here:
```
	/* Do nothing if portal won't return tuples */
	if (portal->tupDesc == NULL)
	{
		/*
		 * For SELECT like queries we delay filling in the tupDesc until after
		 * PortalRunUtility, because we don't know what rows an EXECUTE
		 * command will return. Because we delay setting tupDesc, we also need
		 * to delay setting formats. We do this in a pretty hacky way, by
		 * temporarily setting the portal formats to the passed in formats.
		 * Then once we fill in tupDesc, we call PortalSetResultFormat again
		 * with portal->formats to fill in the final formats value.
		 */
		if (portal->strategy == PORTAL_UTIL_SELECT)
			portal->formats = formats;
		return;
	}
```
`formats` should be copied her because it is create in not  appreciate  memory context.
See how it is done below:
```
	portal->formats = (int16 *)
		MemoryContextAlloc(portal->portalContext,
```

Reply:
Yeah, my patches were actually breaking Bind messages indeed. I sent in a new version that fixes those.

Reply:
Great, thanks for the patch, some of the failed test cases ran successfully
Now a new bug, I added a minimal example to the old_oid.py file

```
python -m old_oid -D postgresql+asyncpg://test:test@pgbouncer_ps:5432/test

.......
    stmt = await self._get_statement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/marat/tmp/docker-pgbouncer/venv/lib/python3.11/site-packages/asyncpg/connection.py", line 429, in _get_statement
    raise exceptions.InternalClientError(
asyncpg.exceptions._base.InternalClientError: could not resolve query result and/or argument types in 6 attempts
```

Briefly what happens:

1) Initial call to prepare and execute
```
LOG:  execute PGBOUNCER_7: CREATE TYPE myenum AS ENUM ('foo', 'bar')
LOG:  execute PGBOUNCER_8: 
         CREATE TABLE foo (
                 bar myenum
         )
```

2) prepare PGBOUNCER_9 as `INSERT INTO foo (bar) VALUES ($1::myenum)`.

3) Somehow the variable `types_with_missing_codecs` is calculated, where the oid '{16621}' is recorded in it
https://github.com/MagicStack/asyncpg/blob/master/asyncpg/connection.py#L401-L435
I understand this is an attempt to calculate the type by oid or something similar

4) Next insert into the table
5) rollback transaction

6) A new transaction is created and the enum type and table are re-created.
```
LOG:  execute PGBOUNCER_7: CREATE TYPE myenum AS ENUM ('foo', 'bar')
LOG:  execute PGBOUNCER_8: 
         CREATE TABLE foo (
                 bar myenum
         )
```

7) Prepare PGBOUNCER_9 as `INSERT INTO foo (bar) VALUES ($1::myenum)` again.

8) The `types_with_missing_codecs` variable is calculated somehow again
https://github.com/MagicStack/asyncpg/blob/master/asyncpg/connection.py#L401-L435

The `types_with_missing_codecs` again contains the oid '{16621}', but it no longer exists!!!
After 6 attempts to perform a search (search for what???) an exception is called


Apparently the oid is stored in prepare_statement, which should be reset/recalculated at any ddl


Tested on:

postgres: commit 36e4419d1f1ef06bba58a28a870aaaa8de73bb46 + patches 
\+ pgbouncer, current branch -- exception


postgres: commit 36e4419d1f1ef06bba58a28a870aaaa8de73bb46 + patches 
\+ pgbouncer, 1.20.0 -- successful

Reply:
I updated my postgres patchset to address this issue there.

Reply:
Great, 29,168 tests passed
There are 8 tests that failed, but they also fail when connecting to postgres directly, without pgbouncer. This doesn't seem to have anything to do with the current changes.

I also missed 2 classes with tests, as there was a call to `SET SESSION search_path=` on each connection

You can see the result [here](https://editor.mergely.com/pscUypwx)